db-engine/src/main/java/com/splicemachine/db/iapi/stats/ColumnStatisticsImpl.java

```diff
@@ -228,10 +233,41 @@ public long selectivity(DataValueDescriptor element) {
         // Use Null Data
         if (element == null || element.isNull())
             return nullCount;
+
         // Frequent Items Sketch?
-        long count = frequenciesSketch.getEstimate(element);
-        if (count>0)
-            return count;
+
+        // When look up predicate value in quantilesSketch, the element's data type
+        // must match with the data type stored the ItemsSketch.hashMap
+        // Otherwise the same value of element in ItemsSketch.hashMap can't be queried out
+        // This handle the situation of string predict compare with DATE/TIME/TIMESTAMP
+        DataValueDescriptor lookUpElement = element;
+        int typeFormatId = dvd.getTypeFormatId();
+        int eTypeFormatId = element.getTypeFormatId();
+        boolean isConverted = false;
+        if (typeFormatId != eTypeFormatId) {
+            if (eTypeFormatId == StoredFormatIds.SQL_CHAR_ID) {
+                try {
+                    switch (typeFormatId) {
+                        case StoredFormatIds.SQL_DATE_ID:
+                            lookUpElement = new SQLDate(element.getDate(new GregorianCalendar()));
+                            break;
+                        case StoredFormatIds.SQL_TIME_ID:
+                            lookUpElement = new SQLTime(element.getTime(null));
+                            break;
+                        case StoredFormatIds.SQL_TIMESTAMP_ID:
+                            lookUpElement = new SQLTimestamp(element.getTimestamp(new GregorianCalendar()));
+                    }
+                    isConverted = true;
+                } catch (StandardException e) {
+                    LOG.warn("Data type conversion failure when looking up frequencies.ItemsSketch", e);
+                }
+            }
+        }
+        if (typeFormatId == eTypeFormatId || isConverted) {
```

Comment:
I agree with Yi.  I did some tests, and estimates when comparing dates with timestamps is currently correct, so no need to bypass the current code.  Maybe just reverse the order of the if statements from:

if (typeFormatId != eTypeFormatId) {
            if (eTypeFormatId == StoredFormatIds.SQL_CHAR_ID) {

... to:

if (eTypeFormatId == StoredFormatIds.SQL_CHAR_ID) {
            if (typeFormatId != eTypeFormatId) {

For the Char test, you could also do: "if (element instanceof SQLChar)" which would handle VARCHAR to since SQLVarchar is a subclass of SQLChar.